README.md
### Mobs Humans
![Mobs Humans' screenshot](screenshot.png)  
**_Adds humans._**

**Version:** 0.4.0  
**Source code's license:** [EUPL v1.2][1] or later, modified zlib-License (cme.lua) see "ZLIB_LICENSE.txt" for details.  
**Textures' license:** [CC BY-SA v3.0][2], [CC BY-SA v4.0][3]

**Dependencies:** default, farming, flowers (found in [Minetest Game][4]), mobs ([Mobs Redo][5])  
**Supported:** [3D Armor][6], [Bonemeal][7]


### Installation

Unzip the archive, rename the folder to mobs_humans and place it in  
../minetest/mods/

If you only want this to be used in a single world, place it in  
../minetest/worlds/WORLD_NAME/worldmods/

GNU+Linux - If you use a system-wide installation place it in  
~/.minetest/mods/

For further information or help see:  
https://wiki.minetest.net/Help:Installing_Mods



[1]: https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863
[2]: https://creativecommons.org/licenses/by-sa/3.0/
[3]: https://creativecommons.org/licenses/by-sa/4.0/
[4]: https://github.com/minetest/minetest_game
[5]: https://forum.minetest.net/viewtopic.php?t=9917
[6]: https://forum.minetest.net/viewtopic.php?t=4654
[7]: https://forum.minetest.net/viewtopic.php?t=16446


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/).


## [Unreleased]

	- No further features planned.



## [0.4.0] - 2019-11-28
### Added

	- Many customization options via Settings -> All Settings -> Mods -> mobs_humans
	- Food drop chance on death.
	- Weapon drop chance on death if killed by a player.
	- Debug chat message (damage math from BlockMen's Creatures MOB-Engine).
	- Support for 3D Armor, Bonemeal.
	- Own mesh model.
	- Weapon draw/sheath.

### Changed

	- Mod in basic mode if 3D Armor not found, can be overriden via Options.
	- Code split into different files.
	- Code improvements, bugfixes, additions, etc.
	- Random values set to realistic chance by default, can be overriden via Options.
	- Default max armor level: 37 (was 10), same as a 3D Armor diamond suit.



## [0.3.0] - 2019-10-17
### Added

	- Support for translations.
	- Option to toggle hovering nametags.

### Changed

	- License changed to EUPL v1.2.
	- mod.conf set to follow MT v5.x specifics.
	- Textures have been optimized (with optipng).
	- Fixed collisionbox.

### Removed

	- Support for MT v0.4.x



## [0.2.2] - 2018-07-27
### Changed

	- Removed redundant conditional check.



## [0.2.1] - 2018-07-27
### Changed

	- Thrown stones will despawn when hitting a node.
	- Minor code improvements.



## [0.2.0] - 2018-07-16
### Added

	- Stone throwing ability.

### Changed

	- Active Objects Count decreased to 1 (was 2)
	- Humans' bones remove timer increased to (min300, max600)secs (was 60, 300)



## [0.2.0-dev] - 2018-07-10
### Added

	- changelog.md

### Changed

	- Health recover function.
	- Bones spawn chance increased to 6/12 (was 4/12)


cme.lua
--[[

	---

	ATTRIBUTION: 'Limit' and 'Calcule Punch Damage' have been taken
	from 'Creatures MOB-Engine (cme)', on_hitted.lua, by
	Copyright (C) 2017 Mob API Developers and Contributors
	Copyright (C) 2015-2016 BlockMen <blockmen2015@gmail.com>

	THIS CODE HAS BEEN MODIFIED TO BE INTEGRATED AS GLOBAL FUNCTIONS
	BY Hamlet.

	See also 'Entity damage mechanism' (../minetest/doc/lua_api.txt)

	---

	= Creatures MOB-Engine (cme) =
	Copyright (C) 2017 Mob API Developers and Contributors
	Copyright (C) 2015-2016 BlockMen <blockmen2015@gmail.com>

	on_hitted.lua

	This software is provided 'as-is', without any express or implied warranty. In no
	event will the authors be held liable for any damages arising from the use of
	this software.

	Permission is granted to anyone to use this software for any purpose, including
	commercial applications, and to alter it and redistribute it freely, subject to the
	following restrictions:

	1. The origin of this software must not be misrepresented; you must not
	claim that you wrote the original software. If you use this software in a
	product, an acknowledgment in the product documentation is required.
	2. Altered source versions must be plainly marked as such, and must not
	be misrepresented as being the original software.
	3. This notice may not be removed or altered from any source distribution.

--]]


-- Limit
mobs_humans.limit = function(value, min, max)
	if value < min then
		return min
	end
	if value > max then
		return max
	end
	return value
end


-- Calcule Punch Damage
mobs_humans.calcPunchDamage = function(obj, actual_interval, tool_caps)
	local damage = 0
	if not tool_caps or not actual_interval then
		return 0
	end
	local my_armor = obj:get_armor_groups() or {}
	for group,_ in pairs(tool_caps.damage_groups) do
		damage = damage + (tool_caps.damage_groups[group] or 0) *
			mobs_humans.limit(actual_interval / tool_caps.full_punch_interval, 0.0, 1.0) *
			((my_armor[group] or 0) / 100.0)
	end
	return damage or 0
end


npc.lua
--[[
	Mobs Humans - Adds human mobs.
	Copyright © 2018, 2019 Hamlet <hamlatmesehub@riseup.net> and contributors.

	Licensed under the EUPL, Version 1.2 or – as soon they will be
	approved by the European Commission – subsequent versions of the
	EUPL (the 'Licence');
	You may not use this work except in compliance with the Licence.
	You may obtain a copy of the Licence at:

	https://joinup.ec.europa.eu/software/page/eupl
	https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863

	Unless required by applicable law or agreed to in writing,
	software distributed under the Licence is distributed on an
	'AS IS' basis,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
	implied.
	See the Licence for the specific language governing permissions
	and limitations under the Licence.

--]]


-- Used for localization

local S = minetest.get_translator('mobs_humans')


--
-- Entity definition
--

mobs:register_mob('mobs_humans:human_npc', {
	nametag = 'Human',
	type = 'npc',
	hp_min = 15,
	hp_max = 20,
	armor = 100,
	walk_velocity = 2,
	run_velocity = 4,
	stand_chance = 50,
	walk_chance = 50,
	jump = true,
	jump_height= 1.1,
	stepheight = 1.1,
	pushable = true,
	view_range = 14,
	damage = 1,
	knock_back = true,
	fear_height = 3,
	fall_damage = true,
	lava_damage = 9999,
	suffocation = true,
	floats = false,
	reach = 4,
	attack_chanche = 12,
	attack_monsters = true,
	attack_animals = false,
	attack_npcs = false,
	attack_players = true,
	group_attack = false,
	attack_type = 'dogfight',
	arrow = 'mobs_humans:stone',
	dogshoot_switch = 1,
	dogshoot_count_max = 9,
	dogshoot_count2_max = 5,
	shoot_interval = 3,
	shoot_offset = 1.7,
	runaway_from = {
		'mobs_banshee:banshee',
		'mobs_ghost_redo:ghost',
		'mobs_others:snow_walker'
	},
	pathfinding = 1,
	makes_footstep_sound = true,
	sounds = {
		attack = 'default_punch',
		shoot_attack = 'mobs_swing'
	},
	drops = mobs_humans.t_Drops,
	visual = 'mesh',
	visual_size = {x = 1, y = 1},
	collisionbox = {-0.3, -1.0, -0.3, 0.3, 0.75, 0.3},
	selectionbox = {-0.3, -1.0, -0.3, 0.3, 0.75, 0.3},
	textures = mobs_humans.t_Skins,
	blood_texture = 'mobs_blood.png',
	mesh = 'mobs_humans_character.b3d',
	animation = {
		stand_start = 0,
		stand_end = 79,
		stand_speed = 30,
		walk_start = 168,
		walk_end = 187,
		walk_speed = 15,
		run_start = 168,
		run_end = 187,
		run_speed = 30,
		punch_start = 189,
		punch_end = 198,
		punch_speed = 30,
		die_start = 162,
		die_end = 166,
		die_speed = 0.8
	},

	on_spawn = function(self, pos)
		if (mobs_humans.b_DynamicMode == true) then
			if (self.nametag == 'Human')
			or (self.class == nil) -- For backward compatibility.
			then
				mobs_humans.OnSpawnDynamic(self)

			end

		else
			if (self.nametag == 'Human')
			or (self.class == nil) -- For backward compatibility.
			then
				mobs_humans.OnSpawnNormal(self)

			end

			mobs_humans.NormalizeStats(self) -- Armor = 100; Damage = 1
		end

		mobs_humans.Nametag(self) -- Set the nametag accordingly to settings.
	end,

	do_punch = function(self, hitter, time_from_last_punch, tool_capabilities,
		direction)

		if (mobs_humans.b_DynamicMode == true) then
			mobs_humans.HitFlag(self) -- Used for experience gain.
			mobs_humans.PlayerHitFlag(self, hitter) -- Used for weapon drop.
		end

		if (mobs_humans.b_Debug == true) then -- Debug mode nametag.
			mobs_humans.NametagDebug(self, hitter, time_from_last_punch,
				tool_capabilities, direction)
		end
	end,

	do_custom = function(self, dtime)
		if (mobs_humans.b_DynamicMode == true) then
			-- Heal if possible.
			mobs_humans.Heal(self, dtime, mobs_humans.t_ALLOWED_STATES,
				mobs_humans.t_HIT_POINTS, mobs_humans.t_HEAL_DELAY,
				self.initial_hp)

			-- Gain experience if possible.
			mobs_humans.Experience(self, dtime)

			-- Draw or sheath the sword if needed.
			mobs_humans.SwordToggle(self)
		end
	end,

	on_rightclick = function(self, clicker)
		mobs_humans.Speak(self, clicker)
	end,

	on_die = function(self, pos)
		mobs_humans.DropBones(self, pos)

		if (mobs_humans.b_DynamicMode == true) then
			mobs_humans.DropWeapon(self, pos)
		end
	end
})


--
-- Entity spawner
--

mobs:spawn({
	name = 'mobs_humans:human_npc',
	nodes = {'group:crumbly'},
	neighbors = {'air'},
	chance = mobs_humans.i_SpawnChance,
	active_object_count = 1,
	min_height = mobs_humans.i_MinSpawnHeight,
	max_height = mobs_humans.i_MaxSpawnHeight
})


-- Spawn Egg

mobs:register_egg('mobs_humans:human_npc', S('Spawn Human NPC'),
	'mobs_humans_icon.png')


-- Alias

mobs:alias_mob('mobs:human_npc', 'mobs_humans:human_npc')


init.lua
--[[
	Mobs Humans - Adds human mobs.
	Copyright © 2018, 2019 Hamlet <hamlatmesehub@riseup.net> and contributors.

	Licensed under the EUPL, Version 1.2 or – as soon they will be
	approved by the European Commission – subsequent versions of the
	EUPL (the 'Licence');
	You may not use this work except in compliance with the Licence.
	You may obtain a copy of the Licence at:

	https://joinup.ec.europa.eu/software/page/eupl
	https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863

	Unless required by applicable law or agreed to in writing,
	software distributed under the Licence is distributed on an
	'AS IS' basis,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
	implied.
	See the Licence for the specific language governing permissions
	and limitations under the Licence.

--]]


--
-- Global mod's namespace
--

mobs_humans = {}


--
-- Global general variables
--

mobs_humans.i_MobDifficulty = minetest.settings:get('mob_difficulty')
if (mobs_humans.i_MobDifficulty == nil) then
	mobs_humans.i_MobDifficulty = 1
end

mobs_humans.i_MinSpawnHeight = minetest.settings:get('mobs_humans_min_spawn_height')
if (mobs_humans.i_MinSpawnHeight == nil) then
	mobs_humans.i_MinSpawnHeight = 1
end

mobs_humans.i_MaxSpawnHeight = minetest.settings:get('mobs_humans_max_spawn_height')
if (mobs_humans.i_MaxSpawnHeight == nil) then
	mobs_humans.i_MaxSpawnHeight = 240
end

mobs_humans.i_SpawnChance = minetest.settings:get('mobs_humans_spawn_chance')
if (mobs_humans.i_SpawnChance == nil) then
	mobs_humans.i_SpawnChance = 60000
end

mobs_humans.b_3D_Armor = minetest.get_modpath('3d_armor')
if (mobs_humans.b_3D_Armor == nil) then
	mobs_humans.b_3D_Armor = false
else
	mobs_humans.b_3D_Armor = true
end

mobs_humans.b_Dynamic = minetest.settings:get_bool('mobs_humans_dynamic')
if (mobs_humans.b_Dynamic == nil) then
	mobs_humans.b_Dynamic = false
end

mobs_humans.b_ShowNametags = minetest.settings:get_bool('mobs_humans_use_nametags')
if (mobs_humans.b_ShowNametags == nil) then
	mobs_humans.b_ShowNametags = false
end

mobs_humans.b_ShowStats = minetest.settings:get_bool('mobs_humans_show_stats')
if (mobs_humans.b_ShowStats == nil) then
	mobs_humans.b_ShowStats = false
end

mobs_humans.b_Debug = minetest.settings:get_bool('mobs_humans_debug')
if (mobs_humans.b_Debug == nil) then
	mobs_humans.b_Debug = false
end

mobs_humans.b_DynamicMode = false
if (mobs_humans.b_3D_Armor == false)
and (mobs_humans.b_Dynamic == true)
then
	mobs_humans.b_DynamicMode = true

elseif (mobs_humans.b_3D_Armor == true)
and (mobs_humans.b_Dynamic == false)
then
	mobs_humans.b_DynamicMode = false

end
--print("Dynamic mode: " .. tostring(mobs_humans.b_DynamicMode))


if (mobs_humans.b_DynamicMode == true) then
	-- Maximum armor level.
	mobs_humans.i_MAX_ARMOR_LEVEL = tonumber(minetest.settings:get('mobs_humans_max_armor'))
	if (mobs_humans.i_MAX_ARMOR_LEVEL == nil) then
		mobs_humans.i_MAX_ARMOR_LEVEL = 37 -- 3D Armor diamond armor suit.
	end

	-- When the mob can heal.
	mobs_humans.t_ALLOWED_STATES = {'stand', 'walk'}

	mobs_humans.t_HEAL_DELAY = 4 -- Seconds that must pass before healing.
	mobs_humans.t_HIT_POINTS = 1 -- Hit points per healing step.

	-- How many fights the mob has to survive to increment its stats.
	mobs_humans.i_FIGHTS_ARMOR = minetest.settings:get('mobs_humans_fights_armor')
	mobs_humans.i_FIGHTS_DAMAGE = minetest.settings:get('mobs_humans_fights_damage')

	if (mobs_humans.i_FIGHTS_ARMOR == nil) then
		mobs_humans.i_FIGHTS_ARMOR = 5
	end

	if (mobs_humans.i_FIGHTS_DAMAGE == nil) then
		mobs_humans.i_FIGHTS_DAMAGE = 10
	end

	-- Whether if armors and weapons levels must be completely random
	-- or based on realistic chance.
	mobs_humans.b_REALISTIC_CHANCE = minetest.settings:get_bool('mobs_humans_realistic_chance')
	if (mobs_humans.b_REALISTIC_CHANCE == nil) then
		mobs_humans.b_REALISTIC_CHANCE = true
	end
end


--
-- Local variables
--

local s_ModPath = minetest.get_modpath('mobs_humans')

-- Used for localization

local S = minetest.get_translator('mobs_humans')


-- Items that can be dropped when the mob dies.
mobs_humans.t_Drops = {
	{name = 'default:apple',			chance = 6, min = 1, max = 2},
	{name = 'default:blueberries',		chance = 6, min = 1, max = 5},
	{name = 'farming:bread',			chance = 12, min = 1, max = 2},
	{name = 'farming:seed_cotton',		chance = 6, min = 1, max = 5},
	{name = 'farming:seed_wheat',		chance = 6, min = 1, max = 5},
	{name = 'flowers:mushroom_brown',	chance = 6, min = 1, max = 3},
	{name = 'flowers:mushroom_red', 	chance = 6, min = 1, max = 3}
}

-- Available skins (../minetest/mods/mobs_humans/textures/)
mobs_humans.t_Skins = {
	{'mobs_humans_female_01.png'},
	{'mobs_humans_female_02.png'},
	{'mobs_humans_male_01.png'},
	{'mobs_humans_male_02.png'}
}


--
-- Load
--

dofile(s_ModPath .. '/cme.lua')
dofile(s_ModPath .. '/functions.lua')
dofile(s_ModPath .. '/node.lua')
dofile(s_ModPath .. '/projectile.lua')
dofile(s_ModPath .. '/animal.lua')
dofile(s_ModPath .. '/monster.lua')
dofile(s_ModPath .. '/npc.lua')


--
-- Dummy entity for previous versions.
--

mobs:register_mob('mobs_humans:human', {
	nametag = 'Human dummy',

	after_activate = function(self)
		self.object:remove()
	end
})


--
-- Minetest engine debug logging
--

local s_LogLevel = minetest.settings:get('debug_log_level')

if (s_LogLevel == nil)
or (s_LogLevel == 'action')
or (s_LogLevel == 'info')
or (s_LogLevel == 'verbose')
then
	s_LogLevel = nil
	minetest.log('action', '[Mod] Mobs Humans [v0.4.0] loaded.')
end


animal.lua
--[[
	Mobs Humans - Adds human mobs.
	Copyright © 2018, 2019 Hamlet <hamlatmesehub@riseup.net> and contributors.

	Licensed under the EUPL, Version 1.2 or – as soon they will be
	approved by the European Commission – subsequent versions of the
	EUPL (the 'Licence');
	You may not use this work except in compliance with the Licence.
	You may obtain a copy of the Licence at:

	https://joinup.ec.europa.eu/software/page/eupl
	https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863

	Unless required by applicable law or agreed to in writing,
	software distributed under the Licence is distributed on an
	'AS IS' basis,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
	implied.
	See the Licence for the specific language governing permissions
	and limitations under the Licence.

--]]


-- Used for localization

local S = minetest.get_translator('mobs_humans')


--
-- Entity definition
--

mobs:register_mob('mobs_humans:human_animal', {
	nametag = 'Human',
	type = 'animal',
	hp_min = 15,
	hp_max = 20,
	armor = 100,
	passive = true,
	walk_velocity = 2,
	run_velocity = 4,
	stand_chance = 50,
	walk_chance = 50,
	jump = true,
	jump_height = 1.1,
	stepheight = 1.1,
	runaway = true,
	pushable = true,
	view_range = 14,
	damage = 1,
	knock_back = true,
	fear_height = 3,
	fall_damage = true,
	suffocation = true,
	floats = false,
	reach = 4,
	attack_chance = 1,
	attack_monsters = true,
	attack_animals = false,
	attack_npcs = false,
	attack_players = true,
	group_attack = true,
	attack_type = 'dogfight',
	arrow = 'mobs_humans:stone',
	dogshoot_switch = 1,
	dogshoot_count_max = 9,
	dogshoot_count2_max = 5,
	shoot_interval = 3,
	shoot_offset = 1.7,
	runaway_from = {
		'mobs_banshee:banshee',
		'mobs_ghost_redo:ghost',
		'mobs_others:snow_walker'
	},
	makes_footstep_sound = true,
	sounds = {
		attack = 'default_punch',
		shoot_attack = 'mobs_swing'
	},
	drops = mobs_humans.t_Drops,
	visual = 'mesh',
	visual_size = {x = 1, y = 1},
	collisionbox = {-0.3, -1.0, -0.3, 0.3, 0.75, 0.3},
	selectionbox = {-0.3, -1.0, -0.3, 0.3, 0.75, 0.3},
	textures = mobs_humans.t_Skins,
	blood_texture = 'mobs_blood.png',
	mesh = 'mobs_humans_character.b3d',
	animation = {
		stand_start = 0,
		stand_end = 79,
		stand_speed = 30,
		walk_start = 168,
		walk_end = 187,
		walk_speed = 15,
		run_start = 168,
		run_end = 187,
		run_speed = 30,
		punch_start = 189,
		punch_end = 198,
		punch_speed = 30,
		die_start = 162,
		die_end = 166,
		die_speed = 0.8
	},

	on_spawn = function(self, pos)
		if (mobs_humans.b_DynamicMode == true) then
			if (self.nametag == 'Human')
			or (self.class == nil) -- For backward compatibility.
			then
				mobs_humans.OnSpawnDynamic(self)

			end

		else
			if (self.nametag == 'Human')
			or (self.class == nil) -- For backward compatibility.
			then
				mobs_humans.OnSpawnNormal(self)

			end

			mobs_humans.NormalizeStats(self) -- Armor = 100; Damage = 1
		end

		mobs_humans.Nametag(self) -- Set the nametag accordingly to settings.
	end,

	do_punch = function(self, hitter, time_from_last_punch, tool_capabilities,
		direction)

		if (mobs_humans.b_DynamicMode == true) then
			mobs_humans.HitFlag(self) -- Used for experience gain.
			mobs_humans.PlayerHitFlag(self, hitter) -- Used for weapon drop.
		end

		if (mobs_humans.b_Debug == true) then -- Debug mode nametag.
			mobs_humans.NametagDebug(self, hitter, time_from_last_punch,
				tool_capabilities, direction)
		end
	end,

	do_custom = function(self, dtime)
		if (mobs_humans.b_DynamicMode == true) then
			-- Heal if possible.
			mobs_humans.Heal(self, dtime, mobs_humans.t_ALLOWED_STATES,
				mobs_humans.t_HIT_POINTS, mobs_humans.t_HEAL_DELAY,
				self.initial_hp)

			-- Gain experience if possible.
			mobs_humans.Experience(self, dtime)

			-- Draw or sheath the sword if needed.
			mobs_humans.SwordToggle(self)
		end
	end,

	on_rightclick = function(self, clicker)
		mobs_humans.Speak(self, clicker)
	end,

	on_die = function(self, pos)
		mobs_humans.DropBones(self, pos)

		if (mobs_humans.b_DynamicMode == true) then
			mobs_humans.DropWeapon(self, pos)
		end
	end
})

--
-- Entity spawner
--

mobs:spawn({
	name = 'mobs_humans:human_animal',
	nodes = {'group:crumbly'},
	neighbors = {'air'},
	chance = mobs_humans.i_SpawnChance,
	active_object_count = 1,
	min_height = mobs_humans.i_MinSpawnHeight,
	max_height = mobs_humans.i_MaxSpawnHeight
})


-- Spawn Egg

mobs:register_egg('mobs_humans:human_animal', S('Spawn Human Animal'),
	'mobs_humans_icon.png')


-- Alias

mobs:alias_mob('mobs:human_animal', 'mobs_humans:human_animal')


monster.lua
--[[
	Mobs Humans - Adds human mobs.
	Copyright © 2018, 2019 Hamlet <hamlatmesehub@riseup.net> and contributors.

	Licensed under the EUPL, Version 1.2 or – as soon they will be
	approved by the European Commission – subsequent versions of the
	EUPL (the 'Licence');
	You may not use this work except in compliance with the Licence.
	You may obtain a copy of the Licence at:

	https://joinup.ec.europa.eu/software/page/eupl
	https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863

	Unless required by applicable law or agreed to in writing,
	software distributed under the Licence is distributed on an
	'AS IS' basis,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
	implied.
	See the Licence for the specific language governing permissions
	and limitations under the Licence.

--]]


-- Used for localization

local S = minetest.get_translator('mobs_humans')


--
-- Entity definition
--

mobs:register_mob('mobs_humans:human_monster', {
	nametag = 'Human',
	type = 'monster',
	hp_min = 15,
	hp_max = 20,
	armor = 100,
	walk_velocity = 2,
	run_velocity = 4,
	stand_chance = 50,
	walk_chance = 50,
	jump = true,
	jump_height = 1.1,
	stepheight = 1.1,
	pushable = true,
	view_range = 14,
	damage = 1,
	knock_back = true,
	fear_height = 3,
	fall_damage = true,
	suffocation = true,
	floats = false,
	reach = 4,
	docile_by_day = true,
	attack_chance = 12,
	attack_monsters = true,
	attack_animals = false,
	attack_npcs = true,
	attack_players = true,
	group_attack = false,
	attack_type = 'dogfight',
	arrow = 'mobs_humans:stone',
	dogshoot_switch = 1,
	dogshoot_count_max = 9,
	dogshoot_count2_max = 5,
	shoot_interval = 3,
	shoot_offset = 1.7,
	runaway_from = {
		'mobs_banshee:banshee',
		'mobs_ghost_redo:ghost',
		'mobs_others:snow_walker'
	},
	pathfinding = 1,
	makes_footstep_sound = true,
	sounds = {
		attack = 'default_punch',
		shoot_attack = 'mobs_swing'
	},
	drops = mobs_humans.t_Drops,
	visual = 'mesh',
	visual_size = {x = 1, y = 1},
	collisionbox = {-0.3, -1.0, -0.3, 0.3, 0.75, 0.3},
	selectionbox = {-0.3, -1.0, -0.3, 0.3, 0.75, 0.3},
	textures = mobs_humans.t_Skins,
	blood_texture = 'mobs_blood.png',
	mesh = 'mobs_humans_character.b3d',
	animation = {
		stand_start = 0,
		stand_end = 79,
		stand_speed = 30,
		walk_start = 168,
		walk_end = 187,
		walk_speed = 15,
		run_start = 168,
		run_end = 187,
		run_speed = 30,
		punch_start = 189,
		punch_end = 198,
		punch_speed = 30,
		die_start = 162,
		die_end = 166,
		die_speed = 0.8
	},

	on_spawn = function(self, pos)
		if (mobs_humans.b_DynamicMode == true) then
			if (self.nametag == 'Human')
			or (self.class == nil) -- For backward compatibility.
			then
				mobs_humans.OnSpawnDynamic(self)

			end

		else
			if (self.nametag == 'Human')
			or (self.class == nil) -- For backward compatibility.
			then
				mobs_humans.OnSpawnNormal(self)

			end

			mobs_humans.NormalizeStats(self) -- Armor = 100; Damage = 1
		end

		mobs_humans.Nametag(self) -- Set the nametag accordingly to settings.
	end,

	do_punch = function(self, hitter, time_from_last_punch, tool_capabilities,
		direction)

		if (mobs_humans.b_DynamicMode == true) then
			mobs_humans.HitFlag(self) -- Used for experience gain.
			mobs_humans.PlayerHitFlag(self, hitter) -- Used for weapon drop.
		end

		if (mobs_humans.b_Debug == true) then -- Debug mode nametag.
			mobs_humans.NametagDebug(self, hitter, time_from_last_punch,
				tool_capabilities, direction)
		end
	end,

	do_custom = function(self, dtime)
		if (mobs_humans.b_DynamicMode == true) then
			-- Heal if possible.
			mobs_humans.Heal(self, dtime, mobs_humans.t_ALLOWED_STATES,
				mobs_humans.t_HIT_POINTS, mobs_humans.t_HEAL_DELAY,
				self.initial_hp)

			-- Gain experience if possible.
			mobs_humans.Experience(self, dtime)

			-- Draw or sheath the sword if needed.
			mobs_humans.SwordToggle(self)
		end
	end,

	on_rightclick = function(self, clicker)
		mobs_humans.Speak(self, clicker)
	end,

	on_die = function(self, pos)
		mobs_humans.DropBones(self, pos)

		if (mobs_humans.b_DynamicMode == true) then
			mobs_humans.DropWeapon(self, pos)
		end
	end
})


--
-- Entity spawner
--

mobs:spawn({
	name = 'mobs_humans:human_monster',
	nodes = {'group:crumbly'},
	neighbors = {'air'},
	chance = mobs_humans.i_SpawnChance,
	active_object_count = 1,
	min_height = mobs_humans.i_MinSpawnHeight,
	max_height = mobs_humans.i_MaxSpawnHeight
})


-- Spawn Egg

mobs:register_egg('mobs_humans:human_monster', S('Spawn Human Monster'),
	'mobs_humans_icon.png')


-- Alias

mobs:alias_mob('mobs:human_monster', 'mobs_humans:human_monster')


projectile.lua
--[[
	Mobs Humans - Adds human mobs.
	Copyright © 2018, 2019 Hamlet <hamlatmesehub@riseup.net> and contributors.

	Licensed under the EUPL, Version 1.2 or – as soon they will be
	approved by the European Commission – subsequent versions of the
	EUPL (the 'Licence');
	You may not use this work except in compliance with the Licence.
	You may obtain a copy of the Licence at:

	https://joinup.ec.europa.eu/software/page/eupl
	https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863

	Unless required by applicable law or agreed to in writing,
	software distributed under the Licence is distributed on an
	'AS IS' basis,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
	implied.
	See the Licence for the specific language governing permissions
	and limitations under the Licence.

--]]


--
-- Arrow entity
--

mobs:register_arrow('mobs_humans:stone', {
	visual = 'sprite',
	visual_size = {x = 0.1, y = 0.1},
	textures = {'default_stone.png'},
	velocity = 18,
	hit_player = function(self, player)
		player:punch(self.object, 1,
			{
				full_punch_interval = 0.1,
				damage_groups = {fleshy = 6},
			}
		)
		self.object:remove()
	end,

	hit_mob = function(self, player)
		player:punch(self.object, 1,
			{
				full_punch_interval = 0.1,
				damage_groups = {fleshy = 6},
			}
		)
		self.object:remove()
	end,

	hit_node = function(self, pos, node)
		self.object:remove()
	end
})


node.lua
--[[
	Mobs Humans - Adds human mobs.
	Copyright © 2018, 2019 Hamlet <hamlatmesehub@riseup.net> and contributors.

	Licensed under the EUPL, Version 1.2 or – as soon they will be
	approved by the European Commission – subsequent versions of the
	EUPL (the 'Licence');
	You may not use this work except in compliance with the Licence.
	You may obtain a copy of the Licence at:

	https://joinup.ec.europa.eu/software/page/eupl
	https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX:32017D0863

	Unless required by applicable law or agreed to in writing,
	software distributed under the Licence is distributed on an
	'AS IS' basis,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
	implied.
	See the Licence for the specific language governing permissions
	and limitations under the Licence.

--]]


-- Used for localization

local S = minetest.get_translator('mobs_humans')


--
-- Local variables
--

local i_MIN_BONES_TIME = 60 -- One minute.
local i_MAX_BONES_TIME = 300 -- Five minutes.


--
-- Custom bones node
--

minetest.register_node('mobs_humans:human_bones', {
	description = S('Human Bones'),
	drawtype = 'mesh',
	mesh = 'mobs_humans_bones_model.obj',
	tiles = {'mobs_humans_bones.png'},
	inventory_image = 'mobs_humans_bones_inv.png',
	wield_image = 'mobs_humans_bones_inv.png',
	selection_box =  {
		type = 'fixed',
			fixed = {
				{-0.5, -0.5, -0.5, 0.5, 0.0, 0.5}
			},
	},
	walkable = false,
	pointable = true,
	diggable = true,
	buildable_to = true,
	paramtype = 'light',
	paramtype2 = 'facedir',
	groups = {dig_immediate = 2, falling_node = 1},
	sounds = default.node_sound_gravel_defaults(),

	on_construct = function(pos)
		mobs_humans.BonesTimer(pos, i_MIN_BONES_TIME, i_MAX_BONES_TIME)
		minetest.check_single_for_falling(pos)
	end,

	on_timer = function(pos, elapsed)
		minetest.get_node_timer(pos):stop()
		minetest.swap_node(pos, {name = 'air'})
	end
})


--
-- Bonemeal support
--

if (minetest.get_modpath("bonemeal") ~= nil) then
	minetest.register_craft({
		output = 'bonemeal:bone 8',
		type = 'shapeless',
		recipe = {'mobs_humans:human_bones'}
	})
end
